File: test_extract_primate_receptors.R
# Unit tests for extract_primate_receptors.R
#
# These tests exercise all pure (non-network) helper functions and mock the
# network-dependent parts so no real HTTP requests are made.
#
# Run with:
#   Rscript -e "testthat::test_file('test_extract_primate_receptors.R')"
# or from within an R session:
#   testthat::test_file("test_extract_primate_receptors.R")

library(testthat)
library(mockery)

# Source the module under test.
# sys.nframe() == 0 is FALSE when sourced, so the CLI block is skipped.
source("extract_primate_receptors.R", local = FALSE)

# ---------------------------------------------------------------------------
# next_link
# ---------------------------------------------------------------------------
test_that("next_link returns NULL for empty string", {
  expect_null(next_link(""))
})

test_that("next_link returns NULL when no 'next' rel is present", {
  header <- '<https://example.com/page1>; rel="prev"'
  expect_null(next_link(header))
})

test_that("next_link returns URL when 'next' rel is present", {
  header <- '<https://rest.uniprot.org/uniprotkb/search?cursor=abc>; rel="next"'
  expect_equal(next_link(header),
               "https://rest.uniprot.org/uniprotkb/search?cursor=abc")
})

test_that("next_link picks 'next' from multiple rels", {
  header <- paste0(
    '<https://example.com/prev>; rel="prev", ',
    '<https://example.com/next>; rel="next"'
  )
  expect_equal(next_link(header), "https://example.com/next")
})

test_that("next_link returns NULL for NULL input", {
  expect_null(next_link(NULL))
})

# ---------------------------------------------------------------------------
# parse_ensembl_ids
# ---------------------------------------------------------------------------
test_that("parse_ensembl_ids returns empty vector for empty string", {
  expect_equal(parse_ensembl_ids(""), character(0))
})

test_that("parse_ensembl_ids returns empty vector for NULL", {
  expect_equal(parse_ensembl_ids(NULL), character(0))
})

test_that("parse_ensembl_ids parses a single ENST ID", {
  expect_equal(parse_ensembl_ids("ENST00000351508;"),
               "ENST00000351508")
})

test_that("parse_ensembl_ids strips isoform annotation", {
  expect_equal(parse_ensembl_ids("ENST00000351508 [isoform_1];"),
               "ENST00000351508")
})

test_that("parse_ensembl_ids parses multiple IDs", {
  ids <- parse_ensembl_ids("ENST00000111111 [isoform_1];ENST00000222222;")
  expect_equal(ids, c("ENST00000111111", "ENST00000222222"))
})

test_that("parse_ensembl_ids accepts ENSG IDs", {
  ids <- parse_ensembl_ids("ENSG00000012345;ENSG00000067890 [v2];")
  expect_equal(ids, c("ENSG00000012345", "ENSG00000067890"))
})

test_that("parse_ensembl_ids accepts mixed ENSG and ENST IDs", {
  ids <- parse_ensembl_ids("ENST00000111111;ENSG00000012345;")
  expect_true("ENST00000111111" %in% ids)
  expect_true("ENSG00000012345" %in% ids)
})

test_that("parse_ensembl_ids deduplicates IDs", {
  ids <- parse_ensembl_ids("ENST00000111111;ENST00000111111;")
  expect_equal(ids, "ENST00000111111")
})

test_that("parse_ensembl_ids ignores non-Ensembl tokens", {
  ids <- parse_ensembl_ids("SOMEOTHER00000;ENST00000111111;")
  expect_equal(ids, "ENST00000111111")
})

# ---------------------------------------------------------------------------
# extract_lineage_ids
# ---------------------------------------------------------------------------
sample_xml <- '<?xml version="1.0" ?>
<TaxaSet>
  <Taxon>
    <TaxId>9606</TaxId>
    <LineageEx>
      <Taxon><TaxId>131567</TaxId></Taxon>
      <Taxon><TaxId>2759</TaxId></Taxon>
      <Taxon><TaxId>9443</TaxId></Taxon>
      <Taxon><TaxId>9526</TaxId></Taxon>
    </LineageEx>
  </Taxon>
</TaxaSet>'

test_that("extract_lineage_ids extracts all taxon IDs", {
  ids <- extract_lineage_ids(sample_xml)
  expect_true(9606L %in% ids)
  expect_true(9443L %in% ids)  # Primates
  expect_true(2759L %in% ids)  # Eukaryota
})

test_that("extract_lineage_ids returns integer(0) for empty string", {
  expect_equal(extract_lineage_ids(""), integer(0))
})

# ---------------------------------------------------------------------------
# is_primate_only
# ---------------------------------------------------------------------------
make_homology <- function(taxon_id, species) {
  list(target = list(
    taxon_id = taxon_id,
    species  = species,
    id       = paste0("ENSG_FAKE_", taxon_id)
  ))
}

test_that("is_primate_only returns FALSE for empty orthologues", {
  expect_false(is_primate_only(list()))
})

test_that("is_primate_only returns TRUE when all orthologues are primates", {
  orthos <- list(
    make_homology(9598L, "pan_troglodytes"),
    make_homology(9544L, "macaca_mulatta")
  )
  stub(is_primate_only, "is_primate_taxon", TRUE)
  expect_true(is_primate_only(orthos))
})

test_that("is_primate_only returns FALSE when a non-primate orthologue is present", {
  orthos <- list(
    make_homology(9598L,  "pan_troglodytes"),
    make_homology(10090L, "mus_musculus")
  )
  stub(is_primate_only, "is_primate_taxon", function(taxon_id) taxon_id != 10090L)
  expect_false(is_primate_only(orthos))
})

test_that("is_primate_only returns FALSE when taxon_id is missing", {
  orthos <- list(list(target = list(species = "unknown_species")))
  expect_false(is_primate_only(orthos))
})

# ---------------------------------------------------------------------------
# build_output_row
# ---------------------------------------------------------------------------
make_entry <- function() {
  list(
    "Entry"          = "P12345",
    "Gene Names"     = "RECEPTOR1 REC1",
    "Protein names"  = "Receptor protein 1"
  )
}

make_orthologues <- function() {
  list(
    list(target = list(species = "pan_troglodytes", id = "ENSPTRG00000001")),
    list(target = list(species = "macaca_mulatta",  id = "ENSMMUG00000001"))
  )
}

test_that("build_output_row captures accession", {
  row <- build_output_row(make_entry(), make_orthologues())
  expect_equal(row[["accession"]], "P12345")
})

test_that("build_output_row captures gene_names", {
  row <- build_output_row(make_entry(), make_orthologues())
  expect_true(grepl("RECEPTOR1", row[["gene_names"]]))
})

test_that("build_output_row counts orthologues correctly", {
  row <- build_output_row(make_entry(), make_orthologues())
  expect_equal(row[["num_primate_orthologues"]], 2L)
})

test_that("build_output_row sorts primate_species alphabetically", {
  row     <- build_output_row(make_entry(), make_orthologues())
  species <- strsplit(row[["primate_species"]], "; ", fixed = TRUE)[[1]]
  expect_equal(species, sort(species))
})

test_that("build_output_row handles empty orthologues", {
  row <- build_output_row(make_entry(), list())
  expect_equal(row[["num_primate_orthologues"]], 0L)
  expect_equal(row[["primate_species"]], "")
})

# ---------------------------------------------------------------------------
# is_primate_taxon (with mocked fetch_tax_xml)
# ---------------------------------------------------------------------------
xml_with_primates <- function(taxon_id) {
  sprintf('<TaxaSet><Taxon>
    <TaxId>%d</TaxId>
    <LineageEx>
      <Taxon><TaxId>131567</TaxId></Taxon>
      <Taxon><TaxId>9443</TaxId></Taxon>
    </LineageEx>
  </Taxon></TaxaSet>', taxon_id)
}

xml_without_primates <- function(taxon_id) {
  sprintf('<TaxaSet><Taxon>
    <TaxId>%d</TaxId>
    <LineageEx>
      <Taxon><TaxId>131567</TaxId></Taxon>
      <Taxon><TaxId>7711</TaxId></Taxon>
    </LineageEx>
  </Taxon></TaxaSet>', taxon_id)
}

# Helper: clear the in-memory primate-status cache between tests
clear_primate_cache <- function() {
  rm(list = ls(.is_primate_cache), envir = .is_primate_cache)
}

test_that("is_primate_taxon returns TRUE for a primate taxon", {
  clear_primate_cache()
  stub(is_primate_taxon, "fetch_tax_xml", xml_with_primates(9598L))
  expect_true(is_primate_taxon(9598L))
})

test_that("is_primate_taxon returns FALSE for a non-primate taxon", {
  clear_primate_cache()
  stub(is_primate_taxon, "fetch_tax_xml", xml_without_primates(10090L))
  expect_false(is_primate_taxon(10090L))
})

test_that("is_primate_taxon caches results (fetch_tax_xml called only once)", {
  clear_primate_cache()
  mock_fetch <- mock(xml_with_primates(9598L))
  stub(is_primate_taxon, "fetch_tax_xml", mock_fetch)
  is_primate_taxon(9598L)
  is_primate_taxon(9598L)
  expect_called(mock_fetch, 1L)
})

File: test_extract_primate_receptors.py
"""
Unit tests for extract_primate_receptors.py

These tests exercise all pure (non-network) helper functions and mock the
network-dependent parts so no real HTTP requests are made.
"""

import importlib
import sys
import os
import types
import unittest
from unittest.mock import MagicMock, patch

# ---------------------------------------------------------------------------
# Import module under test
# ---------------------------------------------------------------------------
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import extract_primate_receptors as epr


class TestNextLink(unittest.TestCase):
    """_next_link helper."""

    def test_returns_none_for_empty_string(self):
        self.assertIsNone(epr._next_link(""))

    def test_returns_none_when_no_next(self):
        header = '<https://example.com/page1>; rel="prev"'
        self.assertIsNone(epr._next_link(header))

    def test_returns_url_for_next(self):
        header = '<https://rest.uniprot.org/uniprotkb/search?cursor=abc>; rel="next"'
        self.assertEqual(
            epr._next_link(header),
            "https://rest.uniprot.org/uniprotkb/search?cursor=abc",
        )

    def test_multiple_rels_picks_next(self):
        header = (
            '<https://example.com/prev>; rel="prev", '
            '<https://example.com/next>; rel="next"'
        )
        self.assertEqual(epr._next_link(header), "https://example.com/next")


class TestParseEnsemblIds(unittest.TestCase):
    """parse_ensembl_ids helper."""

    def test_empty_string_returns_empty_list(self):
        self.assertEqual(epr.parse_ensembl_ids(""), [])

    def test_none_equivalent_empty(self):
        self.assertEqual(epr.parse_ensembl_ids(None), [])  # type: ignore[arg-type]

    def test_single_enst_id(self):
        self.assertEqual(
            epr.parse_ensembl_ids("ENST00000351508;"),
            ["ENST00000351508"],
        )

    def test_enst_with_isoform_annotation(self):
        self.assertEqual(
            epr.parse_ensembl_ids("ENST00000351508 [isoform_1];"),
            ["ENST00000351508"],
        )

    def test_multiple_ids(self):
        ids = epr.parse_ensembl_ids("ENST00000111111 [isoform_1];ENST00000222222;")
        self.assertEqual(ids, ["ENST00000111111", "ENST00000222222"])

    def test_ensg_ids_are_also_accepted(self):
        ids = epr.parse_ensembl_ids("ENSG00000012345;ENSG00000067890 [v2];")
        self.assertEqual(ids, ["ENSG00000012345", "ENSG00000067890"])

    def test_mixed_ensg_and_enst(self):
        ids = epr.parse_ensembl_ids("ENST00000111111;ENSG00000012345;")
        self.assertIn("ENST00000111111", ids)
        self.assertIn("ENSG00000012345", ids)

    def test_deduplication(self):
        ids = epr.parse_ensembl_ids("ENST00000111111;ENST00000111111;")
        self.assertEqual(ids, ["ENST00000111111"])

    def test_non_ensembl_tokens_ignored(self):
        ids = epr.parse_ensembl_ids("SOMEOTHER00000;ENST00000111111;")
        self.assertEqual(ids, ["ENST00000111111"])


class TestExtractLineageIds(unittest.TestCase):
    """_extract_lineage_ids helper."""

    SAMPLE_XML = """<?xml version="1.0" ?>
    <TaxaSet>
      <Taxon>
        <TaxId>9606</TaxId>
        <LineageEx>
          <Taxon><TaxId>131567</TaxId></Taxon>
          <Taxon><TaxId>2759</TaxId></Taxon>
          <Taxon><TaxId>9443</TaxId></Taxon>
          <Taxon><TaxId>9526</TaxId></Taxon>
        </LineageEx>
      </Taxon>
    </TaxaSet>"""

    def test_extracts_all_taxon_ids(self):
        ids = epr._extract_lineage_ids(self.SAMPLE_XML)
        self.assertIn(9606, ids)
        self.assertIn(9443, ids)  # Primates
        self.assertIn(2759, ids)  # Eukaryota

    def test_empty_xml_returns_empty_list(self):
        self.assertEqual(epr._extract_lineage_ids(""), [])


class TestIsPrimateOnlyFilter(unittest.TestCase):
    """is_primate_only filter logic."""

    def _make_homology(self, taxon_id: int, species: str) -> dict:
        return {"target": {"taxon_id": taxon_id, "species": species, "id": f"ENSG_FAKE_{taxon_id}"}}

    def test_empty_orthologues_returns_false(self):
        self.assertFalse(epr.is_primate_only([]))

    def test_all_primates_returns_true(self):
        orthos = [
            self._make_homology(9598, "pan_troglodytes"),
            self._make_homology(9544, "macaca_mulatta"),
        ]
        with patch.object(epr, "is_primate_taxon", return_value=True):
            self.assertTrue(epr.is_primate_only(orthos))

    def test_non_primate_present_returns_false(self):
        orthos = [
            self._make_homology(9598, "pan_troglodytes"),
            self._make_homology(10090, "mus_musculus"),  # mouse
        ]

        def mock_is_primate(taxon_id: int) -> bool:
            return taxon_id != 10090

        with patch.object(epr, "is_primate_taxon", side_effect=mock_is_primate):
            self.assertFalse(epr.is_primate_only(orthos))

    def test_missing_taxon_id_returns_false(self):
        orthos = [{"target": {"species": "unknown_species"}}]
        self.assertFalse(epr.is_primate_only(orthos))


class TestBuildOutputRow(unittest.TestCase):
    """build_output_row assembler."""

    def _entry(self):
        return {
            "Entry": "P12345",
            "Gene Names": "RECEPTOR1 REC1",
            "Protein names": "Receptor protein 1",
        }

    def _orthologues(self):
        return [
            {"target": {"species": "pan_troglodytes", "id": "ENSPTRG00000001"}},
            {"target": {"species": "macaca_mulatta", "id": "ENSMMUG00000001"}},
        ]

    def test_accession_captured(self):
        row = epr.build_output_row(self._entry(), self._orthologues())
        self.assertEqual(row["accession"], "P12345")

    def test_gene_names_captured(self):
        row = epr.build_output_row(self._entry(), self._orthologues())
        self.assertIn("RECEPTOR1", row["gene_names"])

    def test_orthologue_count(self):
        row = epr.build_output_row(self._entry(), self._orthologues())
        self.assertEqual(row["num_primate_orthologues"], 2)

    def test_species_sorted_and_joined(self):
        row = epr.build_output_row(self._entry(), self._orthologues())
        species = row["primate_species"].split("; ")
        self.assertEqual(species, sorted(species))

    def test_empty_orthologues(self):
        row = epr.build_output_row(self._entry(), [])
        self.assertEqual(row["num_primate_orthologues"], 0)
        self.assertEqual(row["primate_species"], "")


class TestIsPrimateTaxon(unittest.TestCase):
    """is_primate_taxon with mocked HTTP."""

    def setUp(self):
        epr._is_primate_cache.clear()

    def _xml_with_primates_lineage(self, taxon_id: int) -> str:
        return f"""<TaxaSet><Taxon>
            <TaxId>{taxon_id}</TaxId>
            <LineageEx>
                <Taxon><TaxId>131567</TaxId></Taxon>
                <Taxon><TaxId>9443</TaxId></Taxon>
            </LineageEx>
        </Taxon></TaxaSet>"""

    def _xml_without_primates_lineage(self, taxon_id: int) -> str:
        return f"""<TaxaSet><Taxon>
            <TaxId>{taxon_id}</TaxId>
            <LineageEx>
                <Taxon><TaxId>131567</TaxId></Taxon>
                <Taxon><TaxId>7711</TaxId></Taxon>
            </LineageEx>
        </Taxon></TaxaSet>"""

    def test_primate_taxon_returns_true(self):
        mock_resp = MagicMock()
        mock_resp.text = self._xml_with_primates_lineage(9598)
        with patch.object(epr, "_get", return_value=mock_resp):
            self.assertTrue(epr.is_primate_taxon(9598))

    def test_non_primate_taxon_returns_false(self):
        mock_resp = MagicMock()
        mock_resp.text = self._xml_without_primates_lineage(10090)
        with patch.object(epr, "_get", return_value=mock_resp):
            self.assertFalse(epr.is_primate_taxon(10090))

    def test_result_is_cached(self):
        mock_resp = MagicMock()
        mock_resp.text = self._xml_with_primates_lineage(9598)
        with patch.object(epr, "_get", return_value=mock_resp) as mock_get:
            epr.is_primate_taxon(9598)
            epr.is_primate_taxon(9598)
            # Should only call _get once (second call hits the cache)
            self.assertEqual(mock_get.call_count, 1)


if __name__ == "__main__":
    unittest.main()
